### ISSUE TITLE
Debug info might be corrupt: Unexpected JDWP Error: 35

### ISSUE THREAD

**Issue by chaokunyang** [janino-compiler/janino#100]
Here is my code to use janino:
```java
public class JaninoUtils {

    public static ClassLoader compile(ClassLoader parentClassLoader, CompileUnit... compileUnits) {
        final Map<String, byte[]> classes = toBytecode(parentClassLoader, compileUnits);
        // Set up a class loader that finds and defined the generated classes.
        return new ByteArrayClassLoader(classes, parentClassLoader);
    }

    public static List<byte[]> compileToBytecode(ClassLoader parentClassLoader, CompileUnit... compileUnits) {
        final Map<String, byte[]> classes = toBytecode(parentClassLoader, compileUnits);
        List<byte[]> byteCodes = new ArrayList<>(compileUnits.length);
        for (CompileUnit unit : compileUnits) {
            String key = unit.pkg.replace(".", "/") + "/" + unit.mainClassName + ".class";
            byteCodes.add(classes.get(key));
        }
        return byteCodes;
    }

    private static Map<String, byte[]> toBytecode(ClassLoader parentClassLoader, CompileUnit... compileUnits) {
        MapResourceFinder sourceFinder = new MapResourceFinder();
        for (CompileUnit unit : compileUnits) {
            String stubFileName = unit.pkg.replace(".", "/") + "/" + unit.mainClassName + ".java";
            sourceFinder.addResource(stubFileName, unit.getCode());
        }

        // Storage for generated bytecode
        final Map<String, byte[]> classes = new HashMap<>();
        // Set up the compiler.
        ClassLoaderIClassLoader iClassLoader = new ClassLoaderIClassLoader(parentClassLoader);
        Compiler compiler = new Compiler(sourceFinder, iClassLoader);
        compiler.setClassFileCreator(new MapResourceCreator(classes));
        compiler.setClassFileFinder(new MapResourceFinder(classes));

        compiler.setDebugSource(true);
        compiler.setDebugLines(true);
        compiler.setDebugVars(true);

        // Compile all sources
        try {
            compiler.compile(sourceFinder.resources().toArray(new Resource[0]));
        } catch (CompileException | IOException e) {
            StringBuilder msgBuilder = new StringBuilder("Compile error: \n");
            for (int i = 0; i < compileUnits.length; i++) {
                CompileUnit unit = compileUnits[i];
                if (i != 0) {
                    msgBuilder.append('\n');
                }
                String qualifiedName = unit.pkg.concat(".").concat(unit.mainClassName);
                msgBuilder.append(qualifiedName).append(":\n");
                msgBuilder.append(CodeFormatter.format(unit.getCode()));
            }
            throw new CodegenException(msgBuilder.toString(), e);
        }

        return classes;
    }

}

public class CompileUnit {
    String pkg;
    String mainClassName;
    private String code;
    private Supplier<String> genCodeFunc;

    public CompileUnit(String pkg, String mainClassName, String code) {
        this.pkg = pkg;
        this.mainClassName = mainClassName;
        this.code = code;
    }

    public CompileUnit(String pkg, String mainClassName, Supplier<String> genCodeFunc) {
        this.pkg = pkg;
        this.mainClassName = mainClassName;
        this.genCodeFunc = genCodeFunc;
    }

    public String getCode() {
        if (code == null) {
            Preconditions.checkNotNull(genCodeFunc);
            code = genCodeFunc.get();
        }
        return code;
    }
}
```

When I debug in idea, it shows `Debug info might be corrupt: Unexpected JDWP Error: 35`. Is there anything that I miss?

**Comment by aunkrig** [comment]
Hey Chaokun,

congrats for submitting the 100th Janino issue! ;-)

Some questions:

* Does the error always occur, or does it depend on the content of the CompileUnits?

* Can you please check *which* of the three debug options causes the error?

* *How* do you debug your CompileUnits? As described on the Janino home page?

* I'm not an IDEA user... could you please, if at all possible, try to reproduce the problem with the ECLIPSE JDT debugger to see if the problem is IDE dependent?


CU Arno

**Comment by chaokunyang** [comment]
Hey aunkrig,
I rerun some tests:
* This error always happen in idea.
* I debug according to Janino home page, got `Variables debug info not available`'
* It turns out that `compiler.setDebugVars(true);`  cause `Debug info might be corrupt: Unexpected JDWP Error:`. But without it,  got `Variables debug info not available`'.
It seems it's  IDE dependent. Can't figure out a solution.

**Comment by aunkrig** [comment]
Are you using `org.codehaus.janino.Compiler` or `org.codehaus.commons.compiler.jdk.Compiler`?

If you're using the former, try to set a system property

```
-DUnitCompiler.disassembleClassFilesToStdout=true
```

and copy the output here.

**Comment by chaokunyang** [comment]
I'm using `org.codehaus.janino.Compiler`.
I set "UnitCompiler.disassembleClassFilesToStdout", but don't get any different out put.
 I'm using janino `3.0.11`.
Here's my code to call janino:
```java
  private static Map<String, byte[]> toBytecode(ClassLoader parentClassLoader, CompileUnit... compileUnits) {
        MapResourceFinder sourceFinder = new MapResourceFinder();
        for (CompileUnit unit : compileUnits) {
            String stubFileName = unit.pkg.replace(".", "/") + "/" + unit.mainClassName + ".java";
            sourceFinder.addResource(stubFileName, unit.getCode());
        }

        System.setProperty("UnitCompiler.disassembleClassFilesToStdout", "true");
        // Storage for generated bytecode
        final Map<String, byte[]> classes = new HashMap<>();
        // Set up the compiler.
        ClassLoaderIClassLoader iClassLoader = new ClassLoaderIClassLoader(parentClassLoader);
        Compiler compiler = new Compiler(sourceFinder, iClassLoader);
        compiler.setClassFileCreator(new MapResourceCreator(classes));
        compiler.setClassFileFinder(new MapResourceFinder(classes));

        // set debug flag to get source file names and line numbers for debug and stacktrace.
        // this is also the default behaviour for javac.
        compiler.setDebugSource(true);
        compiler.setDebugLines(true);

        // Compile all sources
        try {
            compiler.compile(sourceFinder.resources().toArray(new Resource[0]));
        } catch (CompileException | IOException e) {
            StringBuilder msgBuilder = new StringBuilder("Compile error: \n");
            for (int i = 0; i < compileUnits.length; i++) {
                CompileUnit unit = compileUnits[i];
                if (i != 0) {
                    msgBuilder.append('\n');
                }
                String qualifiedName = unit.pkg + "." + unit.mainClassName;
                msgBuilder.append(qualifiedName).append(":\n");
                msgBuilder.append(CodeFormatter.format(unit.getCode()));
            }
            throw new CodegenException(msgBuilder.toString(), e);
        }

        return classes;
    }
```
